Circuitry huge update

arsserpentarium opened the pull request:
1.Fixed pathfinder circuit  …
2.Hypo injector and pump now can be inversed if you nive negative amount.
3.Hypo injector and smoke generator has output pin with amount of stored liquid now.Also they have pulse out pins ,which will be activated after their work.
4.All reagent containers are acidproof now and have pin with self reference(to make it more easy to link them)
5.grabber circuit. Can grab up to 10 tiny to small things in inventory
6.shocker circuit. Will shock referented mob with electricity.
7.advanced locator. Like adjastent, but with adjustable up to 8 tiles radius and possibility to find items by name and desctiption.
8.drone capacity and complexity was raised.
9.Now you can wrench large assembly into place.
10.Adjastent locator have 2 pulse out pins. One will be activated if it find something, And one if doesn't
11.reagent scaner. It's like a beaker. But can give a list of contained reagents.
12.reagent filter. Pump,that can transfer only desired reagents,or all,except unwanted.
13.examiner. returns name,description ,relative coordiinates,distance,amount of reagents
14.fuel cell.Converts chemicals to energy.
15. Internal and external battery sensers. A lot like power monitor in transmission circuit. But without transmission. First returns state of assembly's own battety. Second returns state of any battery in view.
16.2 pins memory circuit. Sometimes it is very useful.
17.pulse demultiplexers. A lot like regular demultiplexer. But it doesn't distributes data. Just sends pulse to given activator pin.
18.list operations. choose element by index. write element by index. delete element by index search desired element in list.  lenght of list.
19.encoder/decoder of refs.To send refs throught EPV2
20.Thrower.Can throw ajastent things at given coordinates.
21.strings: find element in string. explode string by delimeter.
22.Now you finally can clone circuits.You can scan them.You can retrieve program to assembling.You can load it in printer to make copy of circuit with all components ,renames,values and wires.

Atermonera commented:
Did you compile and test your code?

arsserpentarium commented:
Yes,I've tested it. But I've realized, that I 've forgotten to add clone disk to protolathe.

Neerti commented:
```
code/modules/integrated_electronics/core/printer.dm:84:error: PR: undefined type: PR
code/modules/integrated_electronics/core/printer.dm:84:error: /obj/item/device/integrated_electronics/prefab: undefined type path
code/modules/integrated_electronics/core/printer.dm:85:error: PR.program: undefined type: PR.program
code/modules/integrated_electronics/core/printer.dm:232:error: PR: undefined type: PR
code/modules/integrated_electronics/core/printer.dm:232:error: /obj/item/device/integrated_electronics/prefab: undefined type path
code/modules/integrated_electronics/core/printer.dm:233:error: PR.program: undefined type: PR.program
```

arsserpentarium commented:
All issues was fixed. I've forgotten to load dmb

Neerti commented:
Will review this soonish.

Neerti commented on code/modules/integrated_electronics/subtypes/lists.dm:
```diff
@@ -52,12 +52,124 @@
+/obj/item/integrated_circuit/list/search
+	name = "search circuit"
+	desc = "This circuit will give index of desired element in the list."
+	extended_desc = "Search will start at 1 position and will return first matching position"
```
Missing period.

Neerti commented on code/modules/integrated_electronics/subtypes/lists.dm:
```diff
@@ -52,12 +52,124 @@
+
+/obj/item/integrated_circuit/list/at
+	name = "at circuit"
+	desc = "This circuit will pick element from the list by number"
```
Missing period here as well.

Atermonera replied:
`This circuit will pick an element from a list by index.`

Neerti commented on code/modules/integrated_electronics/subtypes/lists.dm:
```diff
@@ -52,12 +52,124 @@
+/obj/item/integrated_circuit/list/at
+	name = "at circuit"
+	desc = "This circuit will pick element from the list by number"
+	extended_desc = "If there is no element with such number, result will be null"
```
Missing period.

Neerti commented on code/modules/integrated_electronics/subtypes/lists.dm:
```diff
@@ -52,12 +52,124 @@
+
+/obj/item/integrated_circuit/list/delete
+	name = "delete circuit"
+	desc = "This circuit will pick element from the list by number"
```
Missing period.

Atermonera replied:
`This circuit will delete the element from a list by index.`

Neerti commented on code/modules/integrated_electronics/subtypes/lists.dm:
```diff
@@ -52,12 +52,124 @@
+/obj/item/integrated_circuit/list/delete
+	name = "delete circuit"
+	desc = "This circuit will pick element from the list by number"
+	extended_desc = "If there is no element with such number, result will be null"
```
Missing period.

Neerti commented on code/modules/integrated_electronics/subtypes/lists.dm:
```diff
@@ -52,12 +52,124 @@
+
+/obj/item/integrated_circuit/list/write
+	name = "write circuit"
+	desc = "This circuit will write element in list with given index"
```
Missing period.

Neerti commented on code/modules/integrated_electronics/subtypes/lists.dm:
```diff
@@ -52,12 +52,124 @@
+/obj/item/integrated_circuit/list/write
+	name = "write circuit"
+	desc = "This circuit will write element in list with given index"
+	extended_desc = "If there is no element with such index,it will give the same list,as before."
```
``list,as before.`` -> ``list, as before.``

Atermonera replied:
And ' index,it will' -> 'index, it will'

Neerti commented on code/modules/integrated_electronics/subtypes/lists.dm:
```diff
@@ -52,12 +52,124 @@
+
+obj/item/integrated_circuit/list/len
+	name = "len circuit"
+	desc = "This circuit will give lenght of the list"
```
Missing period at the end.

Neerti commented on code/modules/integrated_electronics/subtypes/lists.dm:
```diff
@@ -52,12 +52,124 @@
+obj/item/integrated_circuit/list/len
+	name = "len circuit"
+	desc = "This circuit will give lenght of the list"
+	extended_desc = "selfdescriptive"
```
Might be better to leave this null, if it's self descriptive.

Neerti commented on code/modules/integrated_electronics/subtypes/input.dm:
```diff
@@ -431,3 +631,83 @@
+	name = "external battery monitor"
+	desc = "This can help to watch battery state of any device in viev"
+	icon_state = "externalbm"
+	extended_desc = "This circuit will give you values of charge,max charge and percentage of any device or battery in viev"
```
``charge,max charge`` -> ``charge, max charge``, and ``viev`` -> ``view.``

Neerti commented on code/modules/integrated_electronics/subtypes/manipulation.dm:
```diff
@@ -203,4 +202,171 @@
+	name = "shocker circuit"
+	desc = "Used to shock adjastent creatures with electricity."
+	icon_state = "shocker"
+	extended_desc = "The circuit accepts a reference to creature,who needs to be shocked. It can shock target on adjastent tiles.."
```
Excessive period at the end.

Atermonera replied:
`adjacent'

Belsima replied:
No space after the comma either.

Neerti commented on code/modules/integrated_electronics/subtypes/manipulation.dm:
```diff
@@ -203,4 +202,171 @@
+	var/mob/living/M = AM
+	if(!M.Adjacent(T))
+		return //Can't reach
+	M << "<span class='danger'>You feel a sharp shock!</span>"
```
Use ``to_chat()``, please.

arsserpentarium replied:
okay...

Neerti commented on code/modules/integrated_electronics/subtypes/manipulation.dm:
```diff
@@ -203,4 +202,171 @@
+	var/datum/effect/effect/system/spark_spread/s = new /datum/effect/effect/system/spark_spread
+	s.set_up(3, 1, M)
+	s.start()
+	M.Weaken(10)
```
Really would rather not introduce another instant stun.  Perhaps making it more in line with a stunbaton would be better?

arsserpentarium replied:
well, okay... I was thinking about raising coolown throught

Neerti commented on code/modules/integrated_electronics/subtypes/manipulation.dm:
```diff
@@ -203,4 +202,171 @@
+
+/obj/item/integrated_circuit/manipulation/grabber
+	name = "grabber"
+	desc = "circuit with it's own inventory for small/medium items ,used to grab and store things."
```
``circuit with it's own inventory for small/medium items ,used to grab and store things.`` -> ``A circuit with it's own inventory for small/medium items, used to grab and store things.``

Neerti commented on code/modules/integrated_electronics/subtypes/manipulation.dm:
```diff
@@ -203,4 +202,171 @@
+	name = "grabber"
+	desc = "circuit with it's own inventory for small/medium items ,used to grab and store things."
+	icon_state = "grabber"
+	extended_desc = "The circuit accepts a reference to thing to be grabbed. It can store up to 10 things.Modes:1 for grab. 0 for eject the first thing. -1 for eject all"
```
Need a space between ``things.`` and ``Modes:1``, so it looks like ``things. Modes:1``.  Also need a period at the end.

Neerti commented on code/modules/integrated_electronics/subtypes/manipulation.dm:
```diff
@@ -203,4 +202,171 @@
+
+/obj/item/integrated_circuit/manipulation/thrower
+	name = "thrower"
+	desc = "compact launcher to throw things from inside or nearby tiles"
```
Add a capital A at the start or capitalize the C.  Also need a period at the end.

Neerti commented on code/modules/integrated_electronics/subtypes/manipulation.dm:
```diff
@@ -203,4 +202,171 @@
+	desc = "compact launcher to throw things from inside or nearby tiles"
+	extended_desc = "The first and second inputs need to be numbers.  They are coordinates to throw thing at, relative to the machine itself.  \
+	The 'fire' activator will cause the mechanism to attempt to throw thing at the coordinates, if possible.  Note that the \
+	projectile need to be insie the machine or to be on adjastent tiles.And to be up to medium size."
```
``insie the machine or to be on adjastent tiles.And to be up`` -> ``inside the machine, or to be on an adjacent tile, and to be up``

Neerti commented on code/modules/integrated_electronics/subtypes/memory.dm:
```diff
@@ -40,8 +40,15 @@
+
 /obj/item/integrated_circuit/memory/medium
-	name = "memory circuit"
+	name = "Medium memory circuit"
```
Because of BYOND, names should be lowercase if they're not a proper noun, or add a ``\improper`` at the start.  Would rather have it be lowercase to be consistant with the rest of the circuits and items in-game.

Neerti commented on code/modules/integrated_electronics/subtypes/memory.dm:
```diff
@@ -40,8 +40,15 @@
 	activate_pin(2)
 
+/obj/item/integrated_circuit/memory/tiny
+	name = "Small memory circuit"
```
Make the S lowercase for reasons explained below.

Neerti commented on code/modules/integrated_electronics/subtypes/reagents.dm:
```diff
@@ -33,52 +45,141 @@
 	icon_state = "injector"
 	extended_desc = "This autoinjector can push reagents into another container or someone else outside of the machine.  The target \
-	must be adjacent to the machine, and if it is a person, they cannot be wearing thick clothing."
+	must be adjacent to the machine, and if it is a person, they cannot be wearing thick clothing.Negative given amount makes injector suck out reagents."
```
Add a space after the period at ``cannot be wearing thick clothing.``

Neerti commented on code/modules/integrated_electronics/subtypes/reagents.dm:
```diff
@@ -33,52 +45,141 @@
-			sleep(3 SECONDS)
-			if(!L.can_be_injected_by(src))
+
+	if(direc == 1)
```
Would it be possible to trim this code down, since it looks like it was copypasted twice to enable two-directional injection?

arsserpentarium replied:
I'll see about it tomorrow.

Neerti commented on code/modules/integrated_electronics/subtypes/reagents.dm:
```diff
@@ -150,4 +269,92 @@
+	origin_tech = list(TECH_MATERIALS = 4, TECH_ENGINEERING = 2, TECH_DATA = 2, TECH_BIO = 2)
+
+/obj/item/integrated_circuit/reagent/storage/big
+	name = "Big reagent storage"
```
Lowercase this, please.

Neerti commented on code/modules/integrated_electronics/subtypes/reagents.dm:
```diff
@@ -150,4 +269,92 @@
+	name = "reagent scanner"
+	desc = "Stores liquid inside, and away from electrical components.  Can store up to 60u.  On pulse this beaker will send list of contained reagents."
+	icon_state = "reagent_scan"
+	extended_desc = "This is effectively an internal cryo beaker."
```
This description appears to be incorrect and a result of copy-pasting.  Let me know if I'm wrong.

Neerti commented on code/modules/integrated_electronics/subtypes/reagents.dm:
```diff
@@ -150,4 +269,92 @@
+	name = "reagent filter"
+	desc = "Filtering liquids by list of desired or unwanted reagents."
+	icon_state = "reagent_filter"
+	extended_desc = "This is a filter, which will move liquids from the source ref to the target ref. \
```
A space should be added after commas, generally.

Neerti commented on code/modules/integrated_electronics/core/printer.dm:
```diff
@@ -175,6 +205,41 @@
+	if(href_list["print"])
+		switch(href_list["print"])
+			if("load")
+				program = input("Put your code there:", "loading", null, null)
```
This piece is probably gonna be the most contentious change.  I'm really against adding a method to copypaste a code to build a circuit design, since it allows for importing a design that was built outside of the current round, by someone who may not have actually designed the circuit and didn't build it themselves, which is one of the things which ruined NTSL for me (as well as the security issues).  Speaking of security issues, this doesn't appear to be sanitized at all, and even if it was, I'm still paranoid at how this could be used to break things.

arsserpentarium replied:
Yes, that's a point. And that's a point to ignore refs , when it is scanning assembly data.
Because it's a way to master your creations for many rounds. Test it, make it more optimised and perfect.
About sanitize,there is sanity check of program.

Atermonera commented on code/modules/integrated_electronics/core/tools.dm:
```diff
@@ -240,6 +240,99 @@
 
+/obj/item/device/integrated_electronics/analyzer
+	name = "circuit analyzer"
+	desc = "This tool can scan assembly and give code ,that is neeed to recreate this assembly in circuit printer."
```
`This tool can scan an assembly in generate code necessary to recreate it in a circuit printer.`

Atermonera commented on code/modules/integrated_electronics/passive/power.dm:
```diff
@@ -82,6 +82,45 @@
+	name = "fuel cell"
+	desc = "Produces electricity from chemicals"
+	icon_state = "chemical_cell"
+	extended_desc = "This is effectively an internal beaker.But it will consume phoron,slime jelly, welding fuel, carbon,\
```
`It will consume and produce power from` [...] `and blood, in order of decreasing efficiency.`

Atermonera commented on code/modules/integrated_electronics/subtypes/converters.dm:
```diff
@@ -67,6 +67,40 @@
 
+/obj/item/integrated_circuit/converter/refcode
+	name = "reference encoder"
+	desc = "This circuit can convert a reference in encoded string,which can be transferred throught EPV2 circuit"
```
`This circuit can encode a reference into a string, which can then be read by an EPV2 circuit.`

Atermonera commented on code/modules/integrated_electronics/subtypes/converters.dm:
```diff
@@ -67,6 +67,40 @@
+
+/obj/item/integrated_circuit/converter/refdecode
+	name = "reference decoder"
+	desc = "This circuit can convert a encoded refference to actual reference."
```
`reference`

Atermonera commented on code/modules/integrated_electronics/subtypes/converters.dm:
```diff
@@ -168,6 +202,56 @@
+/obj/item/integrated_circuit/converter/findstring
+	name = "find text"
+	desc = "This gives position of sample in the string. Or returns 0."
+	extended_desc = "first pin is string,to be examined.Second pin is sample,which shoul be found. \
```
`The first pin is the string to be examined. The second pin is the sample to be found.`

Atermonera commented on code/modules/integrated_electronics/subtypes/converters.dm:
```diff
@@ -168,6 +202,56 @@
 
+/obj/item/integrated_circuit/converter/findstring
+	name = "find text"
+	desc = "This gives position of sample in the string. Or returns 0."
```
`string, or returns 0.`
Returning 0 from a search function is a bad idea unless you index the searched body from 1. Most code defaults the index for searches to 0, so if you get a 0 returned, your stuff was found at the very beginning.
With this, it'd either be at the beginning or nowhere at all, which is a bad uncertainty. I'd return -1 as the not-found value, because nothing has an element at position -1.

Atermonera replied:
Or byond is a heretic and indexes from 1, disregard the return-value stuff

arsserpentarium replied:
Yes,byond is heretic.

Atermonera commented on code/modules/integrated_electronics/subtypes/converters.dm:
```diff
@@ -168,6 +202,56 @@
+	name = "find text"
+	desc = "This gives position of sample in the string. Or returns 0."
+	extended_desc = "first pin is string,to be examined.Second pin is sample,which shoul be found. \
+	For example, 'eat this burger',' ' will give you position 4.This circuit isn't case sensitive."
```
`position 4. This`

Atermonera commented on code/modules/integrated_electronics/subtypes/converters.dm:
```diff
@@ -168,6 +202,56 @@
+
+/obj/item/integrated_circuit/converter/exploders
+	name = "string exploder"
+	desc = "This splits single string by delimiter into list of strings."
```
`This splits a single string` [...] `into a list of strings.`

Atermonera commented on code/modules/integrated_electronics/subtypes/converters.dm:
```diff
@@ -168,6 +202,56 @@
+/obj/item/integrated_circuit/converter/exploders
+	name = "string exploder"
+	desc = "This splits single string by delimiter into list of strings."
+	extended_desc = "This circuits splits a given string into list of strings, based on the string, and given delimiter. \
```
`This circuit splits a given string into a list of strings based on the string and given delimiter.`

Atermonera commented on code/modules/integrated_electronics/subtypes/input.dm:
```diff
@@ -145,6 +189,53 @@
 
+/obj/item/integrated_circuit/input/examiner
+	name = "examiner"
+	desc = "It' s a little machine vision system. It can return name,description,distance,relative coordinates,\
```
`It's` [...] `It can return the name, description, distance, relative coordinates, total amount of reagents, and maximum amount of reagents of the referenced object.`

Atermonera commented on code/modules/integrated_electronics/subtypes/input.dm:
```diff
@@ -431,3 +631,83 @@
 
+/obj/item/integrated_circuit/input/internalbm
+	name = "internal battery monitor"
+	desc = "This can help to watch the internal battery state"
```
`This monitors the charge level of an internal battery.`

Atermonera commented on code/modules/integrated_electronics/subtypes/input.dm:
```diff
@@ -431,3 +631,83 @@
+	name = "internal battery monitor"
+	desc = "This can help to watch the internal battery state"
+	icon_state = "internalbm"
+	extended_desc = "This circuit will give you values of charge,max charge and percentage of the internal battery by demand"
```
`internal battery on demand.`

arsserpentarium commented:
Fixed all spelling mistakes.
Changed shocker circuit.
Added jk,rs,gated D latches
Added starter circuit(sends pulse after power restores to assembly)

Atermonera commented on code/modules/integrated_electronics/core/helpers.dm:
```diff
@@ -69,4 +69,56 @@
\ No newline at end of file
+	return 0
+
+/obj/item/integrated_circuit/proc/asc2b64(var/S)
```
What do these circuits do?

arsserpentarium replied:
It's no circuits. It's helper functions,which converts string to base 64 and reverse. I was planning to implement it further.

Atermonera replied:
We already have a string-ascii converter, which allows for more characters?

arsserpentarium replied:
string to base 64. https://en.wikipedia.org/wiki/Base64

arsserpentarium commented:
It's no circuits. It's helper functions,which converts string to base 64 and reverse. I was planning to implement it further.

Anewbe commented:
I'm closing this, as I told you. Feel free to recreate it in smaller, separate PRs.

arsserpentarium commented:
111
